daemon/logger/loggerutils/logfile.go

```diff
@@ -813,6 +814,17 @@ func tailFiles(ctx context.Context, files []fileOpener, watcher *logger.LogWatch
 
 		dec.Reset(ra)
 
+		// Wait for the context to be cancelled in a separate goroutine so the we
+		// can close the underlying reader while a the decoder may be waiting for
+		// data. This should trigger an early return from the decoder.
+		ctx, cancel := context.WithCancel(ctx)
+		go func() {
+			<-ctx.Done()
+			if err := ra.Close(); err != nil {
+				log.G(ctx).WithError(err).Debug("Error closing log reader")
+			}
+		}()
```

Comment:
Since our minimum supported Go version is go1.21, we can use [`context.AfterFunc`](https://pkg.go.dev/context#AfterFunc) so we don't need to keep a goroutine parked for the entirety of the tailing operation.
```suggestion
		context.AfterFunc(ctx, func() {
			if err := ra.Close(); err != nil {
				log.G(ctx).WithError(err).Debug("Error closing log reader")
			}
		})

Reply:
I certainly can, but `context.AfterFunc` seems incredibly overkill and still has a parked goroutine.

Reply:
In fact, I think `context.AfterFunc` is entirely the wrong solution here, as it seems the entire point is being able to know when the returned cancel function actually cancels.

Reply:
Huh? Knowing when the context is done without parking a goroutine is the entire rationale for `context.AfterFunc`! https://go.dev/issue/57928

Reply:
`contxt.AfterFunc` is literally parking a goroutine and doing a bunch of other internal state management.

Reply:
https://github.com/golang/go/commit/54d429999c00f277c80b5698fb78c7b71f91aad9

> Add support for contexts that implement an AfterFunc method, which
> can be used to avoid the need to start a new goroutine watching
> the Done channel when propagating cancellation signals.

Reply:
So your suggestion here is to implement an afterFunctioner context?

Reply:
For posterity, we discussed this on a maintainers call and the underlying context implementation has an internal optimization that it can use here such that it won't spin up a goroutine.

Reply:
I've updated this to use context.AfterFunc